src/test/app/Clawback_test.cpp

```diff
@@ -0,0 +1,971 @@
+        env.require(balance(bob, alice["USD"](700)));
+        env.require(balance(alice, bob["USD"](-700)));
+
+        // bob is now the holder and fails to clawback
+        env(claw(bob, alice["USD"](200)), ter(tecNO_PERMISSION));
+        env.close();
+
+        // alice successfully claws back
+        env(claw(alice, bob["USD"](200)));
+        env.close();
+
+        env.require(balance(bob, alice["USD"](500)));
+        env.require(balance(alice, bob["USD"](-500)));
+    }
+
+    void
+    testDeleteDefaultLine(FeatureBitset features)
+    {
+        testcase("Delete default trustline");
+        using namespace test::jtx;
+
+        // If clawback results the trustline to be default,
+        // trustline should be automatically deleted
+        Env env(*this, features);
+        Account alice{"alice"};
+        Account bob{"bob"};
+
+        env.fund(XRP(1000), alice, bob);
+        env.close();
+
+        auto const USD = alice["USD"];
+
+        // alice sets asfAllowClawback
+        env(fset(alice, asfAllowClawback));
+        env.close();
+        env.require(flags(alice, asfAllowClawback));
+
+        // alice issues 1000 USD to bob
+        env.trust(USD(1000), bob);
+        env(pay(alice, bob, USD(1000)));
+        env.close();
+
+        BEAST_EXPECT(ownerCount(env, alice) == 0);
+        BEAST_EXPECT(ownerCount(env, bob) == 1);
+
+        env.require(balance(bob, alice["USD"](1000)));
+        env.require(balance(alice, bob["USD"](-1000)));
+
+        // set limit to default,
+        env(trust(bob, USD(0), 0));
+        env.close();
+
+        BEAST_EXPECT(ownerCount(env, alice) == 0);
+        BEAST_EXPECT(ownerCount(env, bob) == 1);
+
+        // alice claws back full amount from bob, and should also delete
+        // trustline
+        env(claw(alice, bob["USD"](1000)));
+        env.close();
+
+        // bob no longer owns the trustline because it was deleted
+        BEAST_EXPECT(ownerCount(env, alice) == 0);
+        BEAST_EXPECT(ownerCount(env, bob) == 0);
+    }
+
+    void
+    testFrozenLine(FeatureBitset features)
+    {
+        testcase("Frozen trustline");
+        using namespace test::jtx;
+
+        // Claws back from frozen trustline
+        // and the trustline should remain frozen
+        Env env(*this, features);
+        Account alice{"alice"};
+        Account bob{"bob"};
+
+        env.fund(XRP(1000), alice, bob);
+        env.close();
+
+        auto const USD = alice["USD"];
+
+        // alice sets asfAllowClawback
+        env(fset(alice, asfAllowClawback));
+        env.close();
+        env.require(flags(alice, asfAllowClawback));
+
+        // alice issues 1000 USD to bob
+        env.trust(USD(1000), bob);
+        env(pay(alice, bob, USD(1000)));
+        env.close();
+
+        env.require(balance(bob, alice["USD"](1000)));
+        env.require(balance(alice, bob["USD"](-1000)));
+
+        // freeze trustline
+        env(trust(alice, bob["USD"](0), tfSetFreeze));
```

Comment:
okay, to summarize:
An issuer `alice` can create a trust line with the `Clawback` feature enabled. She can then proceed to freeze the trust line unilaterally and clawback all of the funds. Is this right?

Reply:
Yup, `Clawback` should be completely independent of the freeze status of the trustline and should not mutate anything. I put this test in since it won't hurt, even though these are completely unrelated features.

Reply:
okay